Support luna-soc 0.3.x

antoinevg opened the pull request:
This PR relies on: `luna.git HEAD`,  https://github.com/greatscottgadgets/luna-soc/pull/33, https://github.com/greatscottgadgets/luna-soc/pull/38, https://github.com/greatscottgadgets/luna-soc/pull/45 and https://github.com/greatscottgadgets/luna-soc/pull/46

- [x] Fix acm_serial example
- [x] General cleanup
- [x] Test, test, test
  - [x] Linux
  - [x] macOS
  - [x] Windows
- [x] Review Guide

---

# Reviewer Guide

## Introduction

This PR contains extensive changes needed to bring the repository up to date with the luna 0.2 and luna-soc 0.3
releases.

It has been extensively tested on Linux, Mac and Windows according to the following test plan:

- [x] Install Python 3.10 (Facedancer needs 3.10)
- [x] Install Cynthion and all dependencies
- [x] Build assets
  - `cd cynthion.git/cynthion/python`
  - `make distclean assets facedancer`
- [x] Build SoC examples & firmware:
  - `cd cynthion.git/firmware/lunasoc-hal`
  - `cargo build --release --examples`
  - `cd cynthion.git/firmware/moondancer`
  - `cargo build --release --examples`
  - `cargo build --release`
- [x] Plug Cynthion in and run:
  - `cynthion info`
- [x] Run facedancer gateware:
  - `cynthion run facedancer`
- [x] Run firmware unit tests:
  - `cd cynthion.git/firmware/moondancer`
  - `python -m unittest`
- [x] Run firmware examples:
  - `cd cynthion.git/firmware/moondancer`
  - `cargo run --release --example acm_serial`
    - `picocom --imap crcrlf -b 115200 /dev/ttyACM1`
  - `cargo run --release --example usb_hal`
    - `python ./scripts/usb_hal.py`
  - `cargo run --release --example bulk_speed_test`
    - `python ./scripts/bulk_speed_test.py`
- [x] Run facedancer unit tests:
  - `cd facedancer.git/`
  - `python -m test.device`
  - install zadig WinUSB driver for "Stress Test Device" (only on windows)
  - `python -m unittest`
- [x] Run facedancer examples:
  - `python examples/rubber-ducky.py`
  - `python examples/usb-proxy.py`
  - `python examples/ftdi-echo.py`                 (not supported on mac)
  - `python examples/mass-storage.py ./vfat.img`   (not supported on windows

The changes made in this PR can be broken down into the following categories:

1. Changes to gateware because of changes to `luna.git`
2. Changes to gateware because of changes to `luna-soc.git`
3. Changes to the `moondancer-pac` peripheral access crate
4. Changes to the `lunasoc-hal` hardware abstraction layer
5. Changes to `smolusb`
6. Changes to the `moondancer` firmware
7. Miscellaneous changes

## Category Details

### 1. Changes to gateware because of changes to `luna.git`

- `cynthion/python/src/gateware/analyzer/top.py`
- `cynthion/python/src/selftest/gateware.py`

These changes are minor and should be self-explanatory.


### 2. Changes to gateware because of changes to `luna-soc.git`

- `cynthion/python/src/gateware/platform/`
- `cynthion/python/src/gateware/facedancer/advertiser.py`
- `cynthion/python/src/gateware/facedancer/info.py`
- `cynthion/python/src/gateware/facedancer/top.py`

The changes to the platform files are minor and only change how we import `amaranth_boards` so that we won't have to
make changes once `amaranth_boards` gets an official pypi release.

Changes for the Facedancer SoC are more involved consist of:

1. The `advertiser.py` and `info.py` peripherals needed to be ported from lambdasoc to amaranth-soc peripherals:
  * Deriving from wiring.Component instead of lambdasoc.Peripheral
  * Using amaranth wiring for Component signatures
  * Using `amaranth_soc.csr.Register` instead of lambdasoc registers
  * Replacing the bus with `amaranth_soc.csr.Signature`
  * Replacing the use of `Record` for pins with amaranth `PinSignature`
2. The Facedancer SoC in `top.py` needed to be ported from the old lambdasoc/LunaSoC base to amaranth-soc and
luna-soc:
  * The idea of a base SoC class you could derive a new SoC from was retired in favour of the explicit, more
readable approach being moved to by upstream `amaranth-soc` developers.
  * SoC's now also derive from `wiring.Component` and take full responsibility for all elements in the design.
  * SoC's now use a "Provider" class to manage the connections between peripherals and platform resources. (With the
exception of the USB peripherals, we still need to do some work in `luna.git` around Records etc. to enable this.)


### 3. Changes to the `moondancer-pac` peripheral access crate

- `cynthion.git/firmware/moondancer-pac/`

All changes made in this directory are changes in the auto-generated code produced by `luna-soc` and
[`svd2rust`](https://github.com/rust-embedded/svd2rust). It does not require review.


### 4. Changes to the `lunasoc-hal` hardware abstraction layer

- `firmware/lunasoc-hal/examples/`
- `firmware/lunasoc-hal/src/`

Almost all lunasoc-hal changes are solely updates to peripheral register names and layout.

The only functional change is the addition of a new method `set_mode(&mut self, timer::Mode)` in `src/timer.rs`
which exposes new functionality that the lambdasoc timer did not have.


### 5. Changes to `smolusb`

- `firmware/smolusb/src/lib.rs`
- `firmware/smolusb/src/traits.rs`

A rust lint and a new helper function in the `ReadEndpoint` trait that provides a clearer semantic distinction
between initial priming of OUT endpoints vs re-enabling them after a packet has been received.


### 6. Changes to the `moondancer` firmware

- `firmware/moondancer/Cargo.toml`
- `firmware/moondancer/examples/`
- `firmware/moondancer/scripts/`
- `firmware/moondancer/src/`
- `firmware/moondancer/test/`

Dependencies have been bumped to the latest, where possible. Further updates are still held back until we update to
`embedded-hal 1.x` and may (posssibly) also require an update to the minimum supported Rust version.

Two examples that had previously been left out of the Cargo file have now been added: `smollest.rs` and
`interrupts_pac.rs`

The vast majority of changes to rust source files are primarily due to changes to peripheral register names and/or
layout.

The notable exception is a smol change to the bulk_speed_test.rs example that had not previously been updated to
earlier changes introduced by
[25e971f](https://github.com/greatscottgadgets/cynthion/commit/25e971f87de862c634e74a2c2951be41b586ae3f)(lines
267-270)

Changes in the `firmware/moondancer/scripts/` directory are a comment correction and updating the `usb_hal.py` test
scripts for some very old changes made to the behaviour of the `usb_hal.rs` example.

Finally, changes in the `firmware/moondancer/test/` directory are a platform check for Windows as the libusb`
`getManufacturer()` and `getProduct()` API calls always fail with an access error on that platform.


### 7. Miscellaneous changes

- `firmware/memory.x` no longer needs the 32 bytes of bootrom that stopped lambdasoc from freaking out and `mainram`
has been renamed to `blockram` to be in line with upstream conventions.
- `cynthion/python/src/__init__.py` drops the vendored `amaranth_stdio` import as it is no longer required for the
UART peripherals and introduces a vendored `amaranth_board` import so we can do a cleaner import in the platform
definition files.


## Conclusion

There are no docs or `README.md` updates in this PR as they will be submitted as a separate PR after review for this
one is complete.

martinling reviewed (changes requested):
I have reviewed all the code changes here, skipping the generated files, and everything looks reasonable.

While going through the recommended sequence of tests, I ran into two issues when running scripts from Facedancer:

1. When running `python -m test.device`, I noticed that if the host is connected to the TARGET port when starting the script, everything works as expected, but if the TARGET port is disconnected, the script crashes with a timeout error after a few seconds and the Moondancer device stops functioning. With the binaries from the Cynthion 0.1.8 release, the script simply waits for the host to connect. Antoine has identified that this is caused by the EPTRI device generating a continuous stream of bus reset interrupts: https://github.com/greatscottgadgets/luna-soc/issues/44. This is a regression which needs to be fixed before this PR can be merged.

2. While running `python examples/ftdi-echo.py` from the Facedancer repository, the script crashes with a timeout/pipe error shortly after the host opens the emulated serial port. This behaviour is host specific, but Antoine was able to reproduce it, and has determined that it is due to a bug that dates back to luna-soc 0.2.4. As such, although the failure doesn't happen with the 0.1.8 release binaries, it was already broken for new builds, and therefore is not a regression in this PR.

antoinevg commented:
> I have reviewed all the code changes here, skipping the generated files, and everything looks reasonable.
> 
> While going through the recommended sequence of tests, I ran into two issues when running scripts from Facedancer:
> 
>     1. When running `python -m test.device`, I noticed that if the host is connected to the TARGET port when starting the script, everything works as expected, but if the TARGET port is disconnected, the script crashes with a timeout error after a few seconds and the Moondancer device stops functioning. With the binaries from the Cynthion 0.1.8 release, the script simply waits for the host to connect. Antoine has identified that this is caused by the EPTRI device generating a continuous stream of bus reset interrupts: [EPTRI device emits continual bus reset interrupts when unplugged. luna-soc#44](https://github.com/greatscottgadgets/luna-soc/issues/44). This is a regression which needs to be fixed before this PR can be merged.

Fixed in: https://github.com/greatscottgadgets/luna-soc/pull/45

 
>     2. While running `python examples/ftdi-echo.py` from the Facedancer repository, the script crashes with a timeout/pipe error shortly after the host opens the emulated serial port. This behaviour is host specific, but Antoine was able to reproduce it, and has determined that it is due to a bug that dates back to luna-soc 0.2.4. As such, although the failure doesn't happen with the 0.1.8 release binaries, it was already broken for new builds, and therefore is not a regression in this PR.

Fixed by reverting https://github.com/greatscottgadgets/luna-soc/commit/26d6a5fc1ffb666dad086d3e67c13bbd36599a2a  in: https://github.com/greatscottgadgets/luna-soc/pull/46

I'm not particularly overjoyed by this but the underlying problem it tried to work around is inherent to using a design such as `eptri` as a facedancer peripheral (see also: https://github.com/greatscottgadgets/facedancer/issues/157) and I believe it's time to start thinking about a replacement.

martinling reviewed (commented):
I've tested those fixes and they solve both problems.

I think the process from here needs to be:
- Merge https://github.com/greatscottgadgets/luna-soc/pull/45
- Merge https://github.com/greatscottgadgets/luna-soc/pull/46
- Cut a new luna-soc 0.3.1 release.
- Bump the luna-soc dependency in this PR to 0.3.1.
- Merge this PR.
